packages/flutter/lib/src/widgets/spell_check.dart

```diff
@@ -0,0 +1,322 @@
+      String newText, String resultsText, List<SuggestionSpan> results) {
+    final List<SuggestionSpan> correctedSpellCheckResults = <SuggestionSpan>[];
+
+    int spanPointer = 0;
+    int offset = 0;
+    int foundIndex;
+    int spanLength;
+    SuggestionSpan currentSpan;
+    SuggestionSpan adjustedSpan;
+    String currentSpanText;
+    String newSpanText = '';
+    bool currentSpanValid = false;
+    bool foundCurrentSpan = false;
+    RegExp regex;
+
+    // Assumes that the order of spans has not been jumbled for optimization
+    // purposes, and will only search since the previously found span.
+    int searchStart = 0;
+
+    while (spanPointer < results.length) {
+      currentSpan = results[spanPointer];
+      currentSpanText =
+          resultsText.substring(currentSpan.range.start, currentSpan.range.end);
+
+      try {
+        newSpanText = newText.substring(
+            currentSpan.range.start + offset, currentSpan.range.end + offset);
+        currentSpanValid = true;
+      } catch (e) {
+        // currentSpan is invalid and needs to be searched for in newText.
+        currentSpanValid = false;
+      }
+
+      if (currentSpanValid && newSpanText == currentSpanText) {
+        searchStart = currentSpan.range.end + offset;
+        adjustedSpan = SuggestionSpan(
+            TextRange(
+                start: currentSpan.range.start + offset, end: searchStart),
+            currentSpan.suggestions
+        );
+        correctedSpellCheckResults.add(adjustedSpan);
+      } else {
+        regex = RegExp('\\b$currentSpanText\\b');
+        foundIndex = newText.substring(searchStart).indexOf(regex);
+
+        if (foundIndex >= 0) {
+          foundIndex += searchStart;
+          spanLength = currentSpan.range.end - currentSpan.range.start;
+          searchStart = foundIndex + spanLength;
+          adjustedSpan = SuggestionSpan(
+              TextRange(start: foundIndex, end: searchStart),
+              currentSpan.suggestions
+          );
+          offset = foundIndex - currentSpan.range.start;
+
+          correctedSpellCheckResults.add(adjustedSpan);
+        }
+      }
+      spanPointer++;
+    }
+
+    return correctedSpellCheckResults;
+  }
+
+  @override
+  TextSpan buildTextSpanWithSpellCheckSuggestions(
+      TextEditingValue value,
+      bool composingWithinCurrentTextRange,
+      TextStyle? style,
+      TextStyle misspelledTextStyle,
+      SpellCheckResults spellCheckResults) {
+    List<SuggestionSpan>? correctedSpellCheckResults;
+
+    final List<SuggestionSpan> rawSpellCheckResults =
+        spellCheckResults.suggestionSpans;
+    final String spellCheckResultsText = spellCheckResults.spellCheckedText;
+
+    if (spellCheckResultsText != value.text) {
+      correctedSpellCheckResults = correctSpellCheckResults(
+          value.text, spellCheckResultsText, rawSpellCheckResults);
+    } else {
+      correctedSpellCheckResults = rawSpellCheckResults;
+    }
```

Comment:
Nits, can shorten the code as:
```suggestion
    List<SuggestionSpan> spellCheckResults = spellCheckResults.suggestionSpans;
    final String spellCheckResultsText = spellCheckResults.spellCheckedText;
    if (spellCheckResultsText != value.text) {
      spellCheckResults = correctSpellCheckResults(
          value.text, spellCheckResultsText, rawSpellCheckResults);
    } 
```